### ISSUE TITLE
Missing path to assets and strange paths in paths.c

### ISSUE THREAD

**Issue by TLogiviere** [DelusionalLogic/NeoComp#9]
Trying to run the code in master we can observe invalid paths generated from paths.c :
```
assets_add_path(): Using asset path ./assets/
assets_add_path(): Using asset path /home/theo/.config//neocomp/assets/
assets_add_path(): Using asset path /etc/xdg//neocomp/assets/
```
Note the '//' in the last two.
Also, it complains about not being able to find shaders :
`No file named stencil.shader`
Of course, adding `assets_add_path('/etc/neocomp/assets')` in compton.c fix this, but I am not sure you want to do it that way.

**Comment by DelusionalLogic** [comment]
According to POSIX, double slashes aren't a problem, and they make it easier for me since I don't have to worry about whether you have a slash at the end of your XDG env variables.

As for the shader path, You're right, although that should be fixed in the latest version. I moved the config to `/etx/xdg/neocomp` as you would expect

**Comment by TLogiviere** [comment]
Alright, good to know !
Now trying to run latest code :
```
 NeoComp $ git rev-parse HEAD
7feebaf9e0a850f72c69217c0496c8df78489717
```
No more crash but still not quite there :
```
~ $ gdb --args /usr/bin/neocomp --config /dev/null --show-all-xerrors -d :0
GNU gdb (GDB) 8.2
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /usr/bin/neocomp...(no debugging symbols found)...done.
(gdb) run
Starting program: /usr/bin/neocomp --config /dev/null --show-all-xerrors -d :0
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
[     0.00 ] error    2 BadValue     request  138 minor    7 serial     10: "BadValue (integer parameter out of range for operation)"
[New Thread 0x7ffff44b0700 (LWP 2385)]
assets_add_path(): Using asset path ./assets/
assets_add_path(): Using asset path /home/theo/.config//neocomp/assets/
assets_add_path(): Using asset path /etc/xdg//neocomp/assets/
Opengl version 4.5 (Core Profile) Mesa 18.2.6
Xorg Features: 
    Composite: 0.2+
    XFIXES: YES
    DAMAGE: YES
    RENDER: YES
    SHAPE: YES
    RANDR: YES
    GLX: YES
    XINERAMA: YES
    SYNC: YES
[     0.15 ] error    2 BadValue     request  142 minor    3 serial     96: "BadValue (integer parameter out of range for operation)"
Uniforms in shader "stencil"
	Uniform "mvp" has id 0, required 0
	Uniform "flip" has id 2, required 0
	Uniform "tex_scr" has id 3, required 1
Wrongly formatted line "v 0.0 1.0 0.0", ignoring
Wrongly formatted line "v 0.0 0.0 0.0", ignoring
Wrongly formatted line "v 1.0 1.0 0.0", ignoring
Wrongly formatted line "v 1.0 1.0 0.0", ignoring
Wrongly formatted line "v 0.0 0.0 0.0", ignoring
Wrongly formatted line "v 1.0 0.0 0.0", ignoring
Wrongly formatted line "vt 0.0 1.0", ignoring
Wrongly formatted line "vt 0.0 0.0", ignoring
Wrongly formatted line "vt 1.0 1.0", ignoring
Wrongly formatted line "vt 1.0 1.0", ignoring
Wrongly formatted line "vt 0.0 0.0", ignoring
Wrongly formatted line "vt 1.0 0.0", ignoring
Wrongly formatted vector2 def "1,1", ignoring
Wrongly formatted vector2 def "1,1", ignoring
Wrongly formatted vector2 def "1,1", ignoring
Uniform "uvscale" is not defined in shader /etc/xdg//neocomp/assets//downscale.shader
[Thread 0x7ffff44b0700 (LWP 2385) exited]
[Inferior 1 (process 2381) exited with code 01]
(gdb) t a a bt
(gdb)
```
It seems the absence of uvscale definition causes the exit ? But it seems to be defined :
```
File: /etc/xdg/neocomp/assets/downscale.shader
   1   │ #version 1
   2   │ 
   3   │ type downsample
   4   │ vertex simple.vs
   5   │ fragment downscale.fs
   6   │ attrib 0 vertex
   7   │ attrib 1 uv
   8   │ 
   9   │ uniform mvp ignored
  10   │ uniform tex_scr sampler
  11   │ uniform uvscale vec2 1,1
  12   │ uniform pixeluv vec2 1,1
  13   │ uniform flip bool
  14   │ uniform extent vec2 1,1
```
So maybe the BadValue errors ?
I must admit I have no clue of the why and where… Any help to debug this, so I can get the os running without tearing, will be super useful :+1:

**Comment by DelusionalLogic** [comment]
I'm gonna take a guess and say it's a locale issue.

What's your locale set to (the output of the locale command). I'm guessing it something where you use commas as decimal separators. If I'm right you should be able to work around it by setting LC_NUMERIC to C.

The scanf parser was supposed to be temporary, but i never got around to replacing it.

**Comment by TLogiviere** [comment]
Yep, setting LC_NUMERIC to C permitted to launch neocomp but then only the wallpaper and the cursor are visible. Windows are not redrawn but the cursor react to their presence.
So I changed every locale to en_GB.UTF-8 and compiled the newest code `7b73339` but to no further progress.
```
$ locale
LANG=en_GB.UTF-8
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=
```
Output of neocomp until termination with Ctrl+C is :
```
[     0.00 ] error    2 BadValue     request  138 minor    7 serial     10: "BadValue (integer parameter out of range for operation)"
assets_add_path(): Using asset path ./assets/
assets_add_path(): Using asset path /home/theo/.config//neocomp/assets/
assets_add_path(): Using asset path /etc/xdg//neocomp/assets/
Opengl version 4.5 (Core Profile) Mesa 18.2.6
Xorg Features: 
    Composite: 0.2+
    XFIXES: YES
    DAMAGE: YES
    RENDER: YES
    SHAPE: YES
    RANDR: YES
    GLX: YES
    XINERAMA: YES
    SYNC: YES
[     0.02 ] error    2 BadValue     request  142 minor    3 serial     96: "BadValue (integer parameter out of range for operation)"
Uniforms in shader "stencil"
	Uniform "mvp" has id 0, required 0
	Uniform "flip" has id 2, required 0
	Uniform "tex_scr" has id 3, required 1
Uniforms in shader "downsample"
	Uniform "mvp" has id 0, required 0
	Uniform "tex_scr" has id 5, required 1
	Uniform "uvscale" has id 1, required 0
	Uniform "pixeluv" has id 3, required 0
	Uniform "flip" has id 2, required 1
	Uniform "extent" has id 4, required 0
Uniforms in shader "upsample"
	Uniform "mvp" has id 0, required 0
	Uniform "flip" has id 2, required 1
	Uniform "tex_scr" has id 5, required 1
	Uniform "uvscale" has id 1, required 0
	Uniform "pixeluv" has id 3, required 0
	Uniform "extent" has id 4, required 0
Uniforms in shader "colored"
	Uniform "mvp" has id 0, required 0
	Uniform "viewport" has id 3, required 1
	Uniform "opacity" has id 5, required 0
	Uniform "color" has id 4, required 0
Uniforms in shader "global"
	Uniform "mvp" has id 0, required 0
	Uniform "flip" has id 2, required 0
	Uniform "tex_scr" has id 3, required 1
	Uniform "invert" has id 6, required 0
	Uniform "dim" has id 5, required 0
	Uniform "opacity" has id 4, required 0
	Uniform "uvscale" has id 1, required 0
Uniforms in shader "passthough"
	Uniform "mvp" has id 0, required 0
	Uniform "flip" has id 2, required 0
	Uniform "opacity" has id 4, required 0
	Uniform "tex_scr" has id 3, required 1
```
When I blindly change the focused window we get the same but with this line
`[ERR] update_ewmh_active_win: Window with the id 46137350 was not found`
Also, when I blindly open a new window in the same workspace, switch to it and then back to terminate neocomp, we get
```
[     0.73 ] error    3 BadWindow    request    3 minor    0 serial   1295: "BadWindow (invalid Window parameter)"
[     0.90 ] error    3 BadWindow    request    3 minor    0 serial   1496: "BadWindow (invalid Window parameter)"
```
Finaly, when I blindly change the focused workspace we get the same as above but with id 0
`[ERR] update_ewmh_active_win: Window with the id 0 was not found`

With hope this can be useful, but in any case,
Happy new year ! :smiley:

**Comment by DelusionalLogic** [comment]
The last error (window id 0) is expected. I'll fix that at some point, but it does no harm.

I don't actually know what happens when you run without a config.

The other ones seem to point to some problem. Could you send me the output of `xdpyinfo -queryExt | grep opcode` so i can see which extension is returning those errors? I'll have a look when I get some spare time. Thanks for the detailed reports, and happy new year :)

**Comment by aufkrawall** [comment]
Hi there,
when compiling with PKGBUILD on Arch, this happens:
>✗ A bezier won't get 1 for 1 when querying on nonlinear curve
>        By floating equality test on v 1.000000==1.000000 

When skipping the tests via `--nocheck`, it compiles. Then it crashes on start (no config specified):
> $ neocomp
[     0.00 ] error    2 BadValue     request  138 minor    7 serial     10: "BadValue (integer parameter out of range for operation)"
assets_add_path(): Using asset path ./assets/
assets_add_path(): Using asset path /home/USER/.config//neocomp/assets/
assets_add_path(): Using asset path /etc/xdg//neocomp/assets/
Opengl version 4.5 (Core Profile) Mesa 19.0.0-devel (git-e869481ef3)
Xorg Features: 
    Composite: 0.2+
    XFIXES: YES
    DAMAGE: YES
    RENDER: YES
    SHAPE: YES
    RANDR: YES
    GLX: YES
    XINERAMA: YES
    SYNC: YES
[     0.04 ] error    2 BadValue     request  142 minor    3 serial     95: "BadValue (integer parameter out of range for operation)"
neocomp: src/framebuffer.c:20: framebuffer_resetTarget: Zusicherung »framebuffer->gl_fbo != 0« nicht erfüllt.
> Abgebrochen (Speicherabzug geschrieben)

This is with RX 580 & Mesa:
> $ xdpyinfo -queryExt | grep opcode
    BIG-REQUESTS  (opcode: 133)
    Composite  (opcode: 142)
    DAMAGE  (opcode: 143, base event: 91, base error: 152)
    DOUBLE-BUFFER  (opcode: 145, base error: 153)
    DPMS  (opcode: 147)
    DRI2  (opcode: 155, base event: 119)
    DRI3  (opcode: 149)
    GLX  (opcode: 152, base event: 95, base error: 158)
    Generic Event Extension  (opcode: 128)
    MIT-SCREEN-SAVER  (opcode: 144, base event: 92)
    MIT-SHM  (opcode: 130, base event: 65, base error: 128)
    Present  (opcode: 148)
    RANDR  (opcode: 140, base event: 89, base error: 147)
    RECORD  (opcode: 146, base error: 154)
    RENDER  (opcode: 139, base error: 142)
    SECURITY  (opcode: 137, base event: 86, base error: 138)
    SHAPE  (opcode: 129, base event: 64)
    SYNC  (opcode: 134, base event: 83, base error: 134)
    X-Resource  (opcode: 150)
    XC-MISC  (opcode: 136)
    XFIXES  (opcode: 138, base event: 87, base error: 140)
    XFree86-DGA  (opcode: 154, base event: 112, base error: 179)
    XFree86-VidModeExtension  (opcode: 153, base error: 172)
    XINERAMA  (opcode: 141)
    XInputExtension  (opcode: 131, base event: 66, base error: 129)
    XKEYBOARD  (opcode: 135, base event: 85, base error: 137)
    XTEST  (opcode: 132)
>    XVideo  (opcode: 151, base event: 93, base error: 155)

**Comment by TLogiviere** [comment]
Hello !
Then, would you mind providing a default config file ?
Here is the output of `xdpyinfo -queryExt | grep opcode`
```
~ $ xdpyinfo -queryExt | grep opcode
    BIG-REQUESTS  (opcode: 133)
    Composite  (opcode: 142)
    DAMAGE  (opcode: 143, base event: 91, base error: 152)
    DOUBLE-BUFFER  (opcode: 145, base error: 153)
    DPMS  (opcode: 147)
    DRI2  (opcode: 155, base event: 119)
    DRI3  (opcode: 149)
    GLX  (opcode: 152, base event: 95, base error: 158)
    Generic Event Extension  (opcode: 128)
    MIT-SCREEN-SAVER  (opcode: 144, base event: 92)
    MIT-SHM  (opcode: 130, base event: 65, base error: 128)
    Present  (opcode: 148)
    RANDR  (opcode: 140, base event: 89, base error: 147)
    RECORD  (opcode: 146, base error: 154)
    RENDER  (opcode: 139, base error: 142)
    SECURITY  (opcode: 137, base event: 86, base error: 138)
    SHAPE  (opcode: 129, base event: 64)
    SYNC  (opcode: 134, base event: 83, base error: 134)
    X-Resource  (opcode: 150)
    XC-MISC  (opcode: 136)
    XFIXES  (opcode: 138, base event: 87, base error: 140)
    XFree86-DGA  (opcode: 154, base event: 112, base error: 179)
    XFree86-VidModeExtension  (opcode: 153, base error: 172)
    XINERAMA  (opcode: 141)
    XInputExtension  (opcode: 131, base event: 66, base error: 129)
    XKEYBOARD  (opcode: 135, base event: 85, base error: 137)
    XTEST  (opcode: 132)
    XVideo  (opcode: 151, base event: 93, base error: 155)
```
No hurry !

**Comment by DelusionalLogic** [comment]
@aufkrawall Could i get you to file a new issue?

@TLogiviere I should probably work on the config parts again soon (I think there's a lot of config options undocumented, and some unused). Thanks for the followup.

**Comment by DelusionalLogic** [comment]
Sure, here it is:

    # Shadow
    shadow = true;                 # Enabled client-side shadows on windows.
    no-dock-shadow = true;         # Avoid drawing shadows on dock/panel windows.
    no-dnd-shadow = true;          # Don't draw shadows on DND windows.
    clear-shadow = true;           # Zero the part of the shadow's mask behind the
                                   # window. Fix some weirdness with ARGB windows.
    shadow-radius = 12;            # The blur radius for shadows. (default 12)
    shadow-offset-x = -15;         # The left offset for shadows. (default -15)
    shadow-offset-y = -15;         # The top offset for shadows. (default -15)
    shadow-opacity = 0.75;         # The translucency for shadows. (default .75)
    # shadow-red = 0.0;            # Red color value of shadow. (0.0 - 1.0, defaults to 0)
    # shadow-green = 0.0;          # Green color value of shadow. (0.0 - 1.0, defaults to 0)
    # shadow-blue = 0.0;           # Blue color value of shadow. (0.0 - 1.0, defaults to 0)
    shadow-exclude = [ 
        "name ?= \"vis\"",
        "class_g = \"Conky\"",
        "class_g ?= \"notify-osd\"",
        "name ?= \"cairo-dock\"",
        "class_g ?= \"Dockx\"",
        "class_g = 'Peek'"
    ];
                                   # Exclude conditions for shadows.
    #shadow-ignore-shaped = true;   # Avoid drawing shadow on all shaped windows
                                   # (see also: --detect-rounded-corners)

    # Opacity
    #menu-opacity = 0.9;                 # The opacity for menus. (default 1.0)
    inactive-opacity = 1.0;              # Default opacity of inactive windows. (0.0 - 1.0)
    # active-opacity = 1.0;              # Default opacity for active windows. (0.0 - 1.0)
    opacity-fade-time = 300;
    bg-opacity-fade-time = 300;
    dim-fade-time = 200;
    opacity-rule = [
        "100:_NET_WM_STATE[0]:32a = '_NET_WM_STATE_FULLSCREEN'",
        "100:_NET_WM_STATE[1]:32a = '_NET_WM_STATE_FULLSCREEN'",
        "100:_NET_WM_STATE[2]:32a = '_NET_WM_STATE_FULLSCREEN'",
        "100:_NET_WM_STATE[3]:32a = '_NET_WM_STATE_FULLSCREEN'",
        "100:_NET_WM_STATE[4]:32a = '_NET_WM_STATE_FULLSCREEN'",
        "100:_NET_WM_STATE[5]:32a = '_NET_WM_STATE_FULLSCREEN'",
        "100:_NET_WM_STATE[6]:32a = '_NET_WM_STATE_FULLSCREEN'",
    ];
    # inactive-opacity-override = true;  # Let inactive opacity set by 'inactive-opacity' overrides
                                         # value of _NET_WM_OPACITY. Bad choice.
    inactive-dim = 0.70;                 # Dim inactive windows. (0.0 - 1.0)
    blur-background = true;              # Blur background of transparent windows.
                                         # Bad performance with X Render backend.
                                         # GLX backend is preferred.
    blur-level = 4;
    blur-background-frame = true;        # Blur background of opaque windows with transparent
                                         # frames as well.
    blur-background-exclude = [ "window_type = 'dock'", "window_type = 'desktop'", "class_g ?= 'Dockx'", "class_g = 'Peek'"];
                                         # Exclude conditions for background blur.

    # Fading
    fading = true;                                                 # Fade windows during opacity changes.
    fade-delta = 10;                                               # The time between steps in a fade in milliseconds. (default 10).
    fade-in-step = 0.02;                                           # Opacity change between steps while fading in. (default 0.028).
    fade-out-step = 0.02;                                          # Opacity change between steps while fading out. (default 0.03).
    no-fading-openclose = true;                                  # Avoid fade windows in/out when opening/closing.
    fade-exclude = [
        "name ?= 'cairo-dock'",
        "class_g ?= 'Dockx'",
    ]; # Exclude conditions for fading.

    # Other
    backend = "glx"                 # Backend to use: "xrender" or "glx". GLX backend is typically
                                    # much faster but depends on a sane driver.
    mark-wmwin-focused = false;     # Try to detect WM windows and mark them as active.
    mark-ovredir-focused = false;   # Mark all non-WM but override-redirect windows active (e.g. menus).
    use-ewmh-active-win = false;    # Use EWMH _NET_WM_ACTIVE_WINDOW to determine which window is focused
                                    # instead of using FocusIn/Out events. Usually more reliable but
                                    # depends on a EWMH-compliant WM.
    detect-rounded-corners = true;  # Detect rounded corners and treat them as rectangular when --shadow-ignore- shaped is on.
    detect-client-opacity = true;   # Detect _NET_WM_OPACITY on client windows, useful for window
                                    # managers not passing _NET_WM_OPACITY of client windows to frame
                                    # windows.
    refresh-rate = 0;               # For --sw-opti: Specify refresh rate of the screen. 0 for auto.
    vsync = "opengl";               # "none", "drm", "opengl", "opengl-oml", "opengl-swc", "opengl-mswc"
                                    # See man page for more details.
    dbe = false;                    # Enable DBE painting mode. Rarely needed.
    paint-on-overlay = true;        # Painting on X Composite overlay window. Recommended.
    sw-opti = false;                # Limit compton to repaint at most once every 1 / refresh_rate.
                                    # Incompatible with certain VSync methods.
    unredir-if-possible = false;     # Unredirect all windows if a full-screen opaque window is
                                    # detected, to maximize performance for full-screen windows.
    focus-exclude = [ ];            # A list of conditions of windows that should always be considered
                                    # focused.
    detect-transient = true;        # Use WM_TRANSIENT_FOR to group windows, and consider windows in
                                    # the same group focused at the same time.
    detect-client-leader = true;    # Use WM_CLIENT_LEADER to group windows.
    invert-color-include = [ ];     # Conditions for windows to be painted with inverted color.

    xrender-sync = true;
    xrender-sync-fence = true;
    shadow-blacklist = [
        ""
    ];

    # GLX backend                       # GLX backend fine-tune options. See man page for more info.
    glx-no-stencil = false;              # Recommended.
    glx-copy-from-front = false;        # Useful with --glx-swap-method,
    # glx-use-copysubbuffermesa = true; # Recommended if it works. Breaks VSync.
    glx-no-rebind-pixmap = true;        # Recommended if it works.
    glx-swap-method = 2;                # See man page.

    dbus=true;

    # Window type settings
    wintypes:
    {
      tooltip = { fade = true; shadow = false; opacity = 40; focus = true; };
      dock = { fade = false; shadow = false; opacity = 100; focus = true; };
      utility = { fade = false; shadow = false; opacity = 100; focus = true; };
      # fade: Fade the particular type of windows.
      # shadow: Give those windows shadow
      # opacity: Default opacity for the type of windows.
      # focus: Whether to always consider windows of this type focused.
    };

Keep in mind that a lot of the options in here don't do anything, but i keep them around because this was also the config i used for compton, and i use it for benchmark :) With this config compton generally works for me (although it also works for me without any config at all)

**Comment by TLogiviere** [comment]
Hello !
Using your config file, neocomp starts ! :tada: 
But it was a bit slow and resource hungry (5-10% cpu time for neocomp, 10-30% for xorg) I guessed because of some config option as you use it for bench. So I started disabling some and found that when setting `blur-background=false;` I get the same result as before : windows are not drawn.
Also, setting `fading=false;` does not disable fading. After restarting neocomp the effect can still be observed by switching the focused window/workspaces/docks.
Disabling vsync with `vsync=none` does not lighten the burden much and I bet the bad perf is because of blur-background and fading not being disabled.
